integration-tests/src/test/java/org/apache/druid/tests/query/ITSqlQueryTest.java

```diff
@@ -0,0 +1,259 @@
+
+  @Inject
+  IntegrationTestingConfig config;
+
+  interface IExecutable
+  {
+    void execute(String endpoint) throws IOException;
+  }
+
+  private void executeWithRetry(String contentType, IExecutable executable)
+  {
+    executeWithRetry(StringUtils.format("%s/druid/v2/sql/", config.getBrokerUrl()), contentType, executable);
+    executeWithRetry(StringUtils.format("%s/druid/v2/sql/", config.getRouterUrl()), contentType, executable);
+  }
+
+  private void executeWithRetry(String endpoint, String contentType, IExecutable executable)
+  {
+    Throwable lastException = null;
+    for (int i = 1; i <= 5; i++) {
+      LOG.info("Query to %s with Content-Type = %s, tries = %s", endpoint, contentType, i);
+      try {
+        executable.execute(endpoint);
+        return;
+      }
+      catch (IOException e) {
+        // Only catch IOException
+        lastException = e;
+      }
+      try {
+        Thread.sleep(200);
+      }
+      catch (InterruptedException ignored) {
+        break;
+      }
+    }
+    throw new ISE(contentType + " failed after 5 tries, last exception: " + lastException);
+  }
+
+  @Test
+  public void testUnsupportedMediaType()
+  {
+    executeWithRetry(
+        "<EMPTY>",
+        (endpoint) -> {
+          try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
+            HttpPost request = new HttpPost(endpoint);
+            request.addHeader("Content-Type", "application/xml");
+            request.setEntity(new StringEntity("select 1"));
+
+            try (CloseableHttpResponse response = client.execute(request)) {
+              assertEquals(HttpStatus.SC_UNSUPPORTED_MEDIA_TYPE, response.getStatusLine().getStatusCode());
+            }
+          }
+        }
+    );
+  }
+
+  @Test
+  public void testTextPlain()
+  {
+    executeWithRetry(
+        "text/plain",
+        (endpoint) -> {
+          try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
+            HttpPost request = new HttpPost(endpoint);
+            request.addHeader("Content-Type", "text/plain");
+            request.setEntity(new StringEntity("select 1"));
+
+            try (CloseableHttpResponse response = client.execute(request)) {
+              assertEquals(200, response.getStatusLine().getStatusCode());
+
+              HttpEntity entity = response.getEntity();
+              assertNotNull(entity);
+              String responseBody = EntityUtils.toString(entity).trim();
+              assertEquals("[{\"EXPR$0\":1}]", responseBody);
+            }
+          }
+        }
+    );
+  }
+
+  @Test
+  public void testFormURLEncoded()
+  {
+    executeWithRetry(
+        "application/x-www-form-urlencoded",
+        (endpoint) -> {
+          try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
+            HttpPost request = new HttpPost(endpoint);
+            request.addHeader("Content-Type", "application/x-www-form-urlencoded");
```

Comment:
For this test please use something that must be urldecoded. For example `SELECT%20%31` would decode to `SELECT 1`.

Reply:
the following line, which is at line 142, sends queries in URL-encoded way

```java
request.setEntity(new StringEntity(URLEncoder.encode("select 1", StandardCharsets.UTF_8)));
```

I can add one more test which contains illegal encoded queries